pkg/execution/state/redis_state/client.go

```diff
@@ -0,0 +1,193 @@
+package redis_state
+
+import (
+	"github.com/oklog/ulid/v2"
+	"github.com/redis/rueidis"
+)
+
+const StateDefaultKey = "estate"
+const QueueDefaultKey = "queue"
+
+type FunctionRunStateClient struct {
+	kg            RunStateKeyGenerator
+	client        RetriableClient
+	unshardedConn RetriableClient
+}
+
+func (f *FunctionRunStateClient) KeyGenerator() RunStateKeyGenerator {
+	return f.kg
+}
+
+func (f *FunctionRunStateClient) Client(runID ulid.ULID) RetriableClient {
```

Comment:
Sharded clients dynamically decide which instance to use based on a switchover key. For function run state, we use the Run ID to place runs after a specific time and date on a new instance.